### PR TITLE
imp: cleanup verifcation arg code for 23-commitment

### PR DESCRIPTION
**Pull request by colin-axner**
## Description

couldn't help myself :smile: 

closes: #XXXX

---

Before we can merge this PR, please make sure that all the following items have been
checked off. If any of the checklist items are not applicable, please leave them but
write a little note why.

- [ ] Targeted PR against the correct branch (see [CONTRIBUTING.md](https://github.com/cosmos/ibc-go/blob/main/docs/dev/pull-requests.md#pull-request-targeting)).
- [ ] Linked to GitHub issue with discussion and accepted design, OR link to spec that describes this work.
- [ ] Code follows the [module structure standards](https://github.com/cosmos/cosmos-sdk/blob/main/docs/build/building-modules/11-structure.md) and [Go style guide](../docs/dev/go-style-guide.md).
- [ ] Wrote unit and integration [tests](https://github.com/cosmos/ibc-go/blob/main/testing/README.md#ibc-testing-package).
- [ ] Updated relevant documentation (`docs/`).
- [ ] Added relevant `godoc` [comments](https://blog.golang.org/godoc-documenting-go-code).
- [ ] Provide a [conventional commit message](https://github.com/cosmos/ibc-go/blob/main/docs/dev/pull-requests.md#commit-messages) to follow the repository standards.
- [ ] Include a descriptive changelog entry when appropriate. This may be left to the discretion of the PR reviewers. (e.g. chores should be omitted from changelog)
- [ ] Re-reviewed `Files changed` in the GitHub PR explorer.
- [ ] Review `SonarCloud Report` in the comment section below once CI passes.

### CODE REVIEW

File: modules/core/23-commitment/types/merkle.go [hunk 1]
```diff
@@ -234,16 +226,8 @@ func (proof *MerkleProof) Empty() bool {
 	return proof == nil || proto.Equal(proof, blankMerkleProof) || proto.Equal(proof, blankProofOps)
```
Comment (colin-axner) [inline modules/core/23-commitment/types/merkle.go:226]: `blankProofOps` check? This can be deleted right?? I don't understand the context of when one could provide that type as `MerkleProof`?

File: modules/core/23-commitment/types/merkle.go [hunk 2]
```diff
@@ -234,16 +226,8 @@ func (proof *MerkleProof) Empty() bool {
...
-	if proof.Empty() {
-		return ErrInvalidProof
-	}
-	return nil
-}
-
 // validateVerificationArgs verifies the proof arguments are valid
-func (proof MerkleProof) validateVerificationArgs(specs []*ics23.ProofSpec, root exported.Root) error {
+func validateVerificationArgs(proof MerkleProof, path v2.MerklePath, specs []*ics23.ProofSpec, root exported.Root) error {
 	if proof.Empty() {
```
Comment (colin-axner) [inline modules/core/23-commitment/types/merkle.go:231]: maybe this should be `if proof == nil || proof.GetProofs() == nil`. I don't feel like the `proof.Empty()` is very effective and creates extra space in this file

File: modules/core/23-commitment/types/merkle_test.go [hunk 3]
```diff
@@ -9,7 +9,7 @@ import (
 	storetypes "cosmossdk.io/store/types"
 
 	"github.com/cosmos/ibc-go/v9/modules/core/23-commitment/types"
-	"github.com/cosmos/ibc-go/v9/modules/core/23-commitment/types/v2"
+	v2 "github.com/cosmos/ibc-go/v9/modules/core/23-commitment/types/v2"
```
Comment (colin-axner) [inline modules/core/23-commitment/types/merkle_test.go:12]:
```suggestion
	"github.com/cosmos/ibc-go/v9/modules/core/23-commitment/types/v2"
```

File: modules/core/23-commitment/types/merkle_test.go [hunk 4]
```diff
@@ -27,9 +27,6 @@ func (suite *MerkleTestSuite) TestVerifyMembership() {
 	proof, err := types.ConvertProofs(res.ProofOps)
 	require.NoError(suite.T(), err)
 
-	suite.Require().NoError(proof.ValidateBasic())
```
Comment (AdityaSripal) [inline modules/core/23-commitment/types/merkle_test.go:30]: Why were these removed?

Comment (colin-axner) [reply modules/core/23-commitment/types/merkle_test.go:30]: This is setup code. `proof.ValidateBasic()` was equivalent to `proof.Empty()`. These tests are testing `TestVerifyMembership` and `TestVerifyNonMembership` not `ConvertProofs`. I could add back a check that the returned proof is not empty, but it's just going to see if the setup code worked correctly

File: modules/core/23-commitment/types/merkle.go (hunk 1 above)
Comment (damiannolan) [reply modules/core/23-commitment/types/merkle.go:226]: Agree, doesn't seem to make sense to me

File: modules/core/23-commitment/types/merkle.go (hunk 2 above)
Comment (damiannolan) [reply modules/core/23-commitment/types/merkle.go:231]: Support this 👍🏻

File: modules/core/23-commitment/types/merkle.go [hunk 5]
```diff
@@ -253,9 +237,11 @@ func (proof MerkleProof) validateVerificationArgs(specs []*ics23.ProofSpec, root
...
 
 	if len(specs) != len(proof.Proofs) {
-		return errorsmod.Wrapf(ErrInvalidMerkleProof,
-			"length of specs: %d not equal to length of proof: %d",
-			len(specs), len(proof.Proofs))
+		return errorsmod.Wrapf(ErrInvalidMerkleProof, "length of specs: %d not equal to length of proof: %d", len(specs), len(proof.Proofs))
+	}
+
+	if len(path.KeyPath) != len(specs) {
+		return errorsmod.Wrapf(ErrInvalidProof, "path length %d not same as proof %d", len(path.KeyPath), len(specs))
```
Comment (damiannolan) [inline modules/core/23-commitment/types/merkle.go:244]: Maybe we could indicate in godoc why this is. e.g. key path of length > 1 is expected to be a chained proof.

Every path element inside `path.KeyPath` is expected to have a corresponding proofSpec, right? In the normal case first we calculate subroots for inclusion/non-inclusion and then inclusion of the subroot in commitment root.

Only in the case of `path.KeyPath == 1`, would we be able to do a single inclusion or non-inclusion proof.

File: modules/core/23-commitment/types/merkle.go (hunk 2 above)
Comment (damiannolan) [reply modules/core/23-commitment/types/merkle.go:231]: Could also just call this `Validate(specs)` as method on merkle proof and move root nil checks to where its being used.

This is also totally fine, and happy to move along with it! 🙏🏻

**Review by damiannolan — APPROVED** [review]
Nice clean up!

File: modules/core/23-commitment/types/merkle.go (hunk 5 above)
Comment (colin-axner) [reply modules/core/23-commitment/types/merkle.go:244]: Correct. This code is expecting to receive a list of keys which corresponds to different trees (which is why it must be a list of specs, 1 per tree)

File: modules/core/23-commitment/types/merkle.go (hunk 2 above)
Comment (colin-axner) [reply modules/core/23-commitment/types/merkle.go]: made the change, but left off the `Validate(specs)` suggestion. Happy to approve a followup, but wasn't exactly sure what changes you had in mind

File: modules/core/23-commitment/types/merkle.go [hunk 6]
```diff
@@ -3,14 +3,11 @@ package types
...
 
-	"github.com/cosmos/gogoproto/proto"
 	ics23 "github.com/cosmos/ics23/go"
 
 	errorsmod "cosmossdk.io/errors"
 
-	cmtcrypto "github.com/cometbft/cometbft/proto/tendermint/crypto"
-
-	"github.com/cosmos/ibc-go/v9/modules/core/23-commitment/types/v2"
+	v2 "github.com/cosmos/ibc-go/v9/modules/core/23-commitment/types/v2"
```
Comment (colin-axner) [inline modules/core/23-commitment/types/merkle.go:10]:
```suggestion
	"github.com/cosmos/ibc-go/v9/modules/core/23-commitment/types/v2"
```

**Review by DimitrisJim — APPROVED** [review]
